ONSdigital/eq-questionnaire-runner
# Update section progress on supplementary data change

### What is the context of this PR?
This PR Fixes two issues that can occur if supplementary data changes when a questionnaire is partially complete

1) If a question depends on the length of a supplementary list, and a new supplementary dataset is used with a list item count that opens up the question, previously the section containing the question would remain completed. It now reverts to in progress. This is true when the supplementary dataset has added or removed items, both will work.

2) If a section contains a list collector with repeating blocks depending on supplementary data. If the supplementary data changes and a list item is removed, the section will revert to in progress.

Note that it does not fix the case where a list item is added and the repeating blocks need to be revisited.

### How to review
Check the changes to the questionnaire store updater make sense. A location wasn't accessible from the session endpoint so it was made optional. And since all occurrences of removing list item data were following it with a call to capture dependencies for the list, this has been moved into the remove list item data method itself.

Additionally the method to capture dependencies for a list has been moved from question to the questionnaire store updater, and the questionnaire store updater is now used to remove list items from the list store when supplementary data changes, rather than doing it by hand so that the dependencies are correctly captured.

### How to test functionally
Run the mock SDS locally on [the branch with an additional dataset](https://github.com/ONSdigital/eq-runner-mock-sds/pull/3)
Use the `test_supplementary_data` schema, and try swapping between the three datasets while clicking the use previous response ID button.

- If using the first supplementary dataset with 2 items, if you complete the production targets section and relaunch the survey with the second supplementary dataset with 3 items, the section should become incomplete.

- If using the second supplementary dataset with 3 items, if you complete the sales targets section, then relaunch the survey using the first supplementary dataset which has 2 items, you should see the section revert to incomplete

- If the product details section is completed whilst using the second supplementary dataset, when moving to the first supplementary dataset which contains fewer items, the section should revert to in progress and force you to reconfirm the calculated summaries within.

- (Note the reverse, swapping to the dataset with more items, does not update completeness of the section, this will be addressed in a follow on card)

### Checklist
